File: apps/iosApp/Sources/ViewModels/ProfileViewModel.swift
import Foundation
import SharedKit

enum ProfileUiState {
    case loading
    case error(String)
    case loaded([Profile])
}

@MainActor
final class ProfileViewModel: ObservableObject {
    @Published private(set) var uiState: ProfileUiState = .loading
    @Published private(set) var isCreating = false

    private let profileRepository: ProfileRepository
    private let activeProfileStore: ActiveProfileStore

    init(profileRepository: ProfileRepository, activeProfileStore: ActiveProfileStore) {
        self.profileRepository = profileRepository
        self.activeProfileStore = activeProfileStore
        load()
    }

    func load() {
        Task {
            uiState = .loading
            do {
                let profiles = try await ThrowingWrappersKt.listProfilesOrThrow(profileRepository)
                uiState = .loaded(profiles)
            } catch {
                uiState = .error(readableMessage(error))
            }
        }
    }

    func selectProfile(_ profileId: String) {
        activeProfileStore.activeProfileId = profileId
    }

    func createProfile(name: String, onCreated: @escaping (String) -> Void) {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        isCreating = true
        Task {
            do {
                let command = CreateProfileCommand(name: trimmed, kind: .standard, language: "en", subtitleLanguage: nil, avatarKey: nil)
                let profile = try await ThrowingWrappersKt.createProfileOrThrow(profileRepository, command: command)
                activeProfileStore.activeProfileId = profile.id
                onCreated(profile.id)
            } catch {
                uiState = .error(readableMessage(error))
            }
            isCreating = false
        }
    }
}

File: apps/androidApp/src/main/kotlin/com/streaming/platform/android/ui/ProfileViewModel.kt
package com.streaming.platform.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.streaming.platform.android.auth.ActiveProfileHolder
import com.streaming.platform.profile.CreateProfileCommand
import com.streaming.platform.profile.Profile
import com.streaming.platform.profile.ProfileRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface ProfileUiState {
    data object Loading : ProfileUiState
    data class Error(val message: String) : ProfileUiState
    data class Loaded(val profiles: List<Profile>) : ProfileUiState
}

class ProfileViewModel(
    private val profileRepository: ProfileRepository,
    private val activeProfileHolder: ActiveProfileHolder,
) : ViewModel() {
    private val _uiState = MutableStateFlow<ProfileUiState>(ProfileUiState.Loading)
    val uiState: StateFlow<ProfileUiState> = _uiState.asStateFlow()

    private val _isCreating = MutableStateFlow(false)
    val isCreating: StateFlow<Boolean> = _isCreating.asStateFlow()

    val activeProfileId: StateFlow<String?> = activeProfileHolder.activeProfileId

    init { load() }

    fun load() {
        viewModelScope.launch {
            _uiState.value = ProfileUiState.Loading
            profileRepository.listProfiles().fold(
                onSuccess = { _uiState.value = ProfileUiState.Loaded(it) },
                onFailure = { _uiState.value = ProfileUiState.Error(it.readableMessage()) },
            )
        }
    }

    fun selectProfile(profileId: String) {
        activeProfileHolder.set(profileId)
    }

    fun createProfile(name: String, onCreated: (String) -> Unit) {
        if (name.isBlank()) return
        _isCreating.value = true
        viewModelScope.launch {
            profileRepository.createProfile(CreateProfileCommand(name = name.trim())).fold(
                onSuccess = { profile ->
                    activeProfileHolder.set(profile.id)
                    onCreated(profile.id)
                },
                onFailure = { _uiState.value = ProfileUiState.Error(it.readableMessage()) },
            )
            _isCreating.value = false
        }
    }
}
